crates/update-notifier/src/lib.rs

```diff
@@ -0,0 +1,216 @@
+use std::time::Duration;
+
+use colored::*;
+use reqwest::Error as ReqwestError;
+use semver::{Error as SemVerError, Version};
+use serde::{Deserialize, Serialize};
+use serde_json::Error as SerdeError;
+use thiserror::Error as ThisError;
+use tiny_gradient::{GradientStr, RGB};
+
+mod fetch;
+mod ui;
+mod utils;
+
+// default interval to check for new updates (one per day)
+const DEFAULT_INTERVAL: Duration = Duration::from_secs(60 * 60 * 24);
+
+#[derive(Debug)]
+pub struct UpdateNotifier {
+    config: UpdateNotifierConfig,
+    package: String,
+    tag: Option<String>,
+    interval: Duration,
+}
+
+// This portion of the config is serialized and persisted to disk
+#[derive(Serialize, Deserialize, Debug, Clone)]
+struct UpdateNotifierConfig {
+    current_version: String,
+    latest_version: Option<String>,
+    last_checked: Option<u128>,
+}
+
+#[derive(ThisError, Debug)]
+pub enum UpdateNotifierError {
+    #[error("Failed to fetch latest version")]
+    FetchError(#[from] ReqwestError),
+    #[error("Failed to parse version")]
+    ParseError(#[from] SemVerError),
+    #[error("Failed to parse JSON")]
+    JsonError(#[from] SerdeError),
+    #[error("Failed to write to terminal")]
+    DisplayLengthError(#[from] ui::utils::GetDisplayLengthError),
+    #[error("Unknown Error")]
+    Internal(),
+}
+
+impl UpdateNotifier {
+    pub fn new(package: String, tag: Option<String>, interval: Option<Duration>) -> Self {
+        let tmp = utils::get_config_path();
+
+        // ignore if it doesn't exist, we don't care
+        if tmp.try_exists().unwrap_or(false) {
+            let file = std::fs::File::open(tmp);
+            let file = match file {
```

Comment:
You could technically do this with `unwrap_or_else` but that's more a stylistic decision.